grpc-api-client/src/main/java/io/deephaven/grpc_api_client/table/BarrageSourcedTable.java

```diff
@@ -0,0 +1,1419 @@
+                            }
+                        }
+                    }
+                }
+            }
+
+            modifiedColumnSet.clear();
+            for (int ii = 0, colIdx = update.modColumns.nextSetBit(0); ii < update.modColumnData.length; ++ii, colIdx = update.modColumns.nextSetBit(colIdx + 1)) {
+                final BarrageMessage.ModColumnData column = update.modColumnData[ii];
+                modifiedColumnSet.setColumnWithIndex(colIdx);
+
+                if (column.rowsIncluded.empty()) {
+                    continue;
+                }
+
+                try (final Index.Iterator outerIter = column.rowsIncluded.iterator();
+                     final WritableLongChunk<Attributes.KeyIndices> keys = WritableLongChunk.makeWritableChunk(column.rowsIncluded.intSize())) {
+                    for (int i = 0; i < keys.size(); ++i) {
+                        final long outerKey = outerIter.nextLong();
+                        keys.set(i, redirectionIndex.get(outerKey));
+                        Assert.notEquals(keys.get(i), "keys[i]", Index.NULL_KEY, "Index.NULL_KEY");
+                    }
+
+                    try (final WritableChunkSink.FillFromContext ctxt = destSources[colIdx].makeFillFromContext(keys.size())) {
+                        destSources[colIdx].fillFromChunkUnordered(ctxt, column.data, keys);
+                    }
+                }
+            }
+
+            beginLog(LogLevel.DEBUG).append(": Populated: ").append(populatedRows).append(", Index=").append(getIndex()).endl();
+
+            if (REPLICATED_TABLE_DEBUG) {
+                Assert.assertion(isViewPort || currentIndex.subsetOf(populatedRows), "isViewPort || currentIndex.subsetOf(populatedRows)");
+            }
+
+            processedDelta = update.lastSeq;
+            notifyOnRowsPopulated(dataOnHand);
+
+            if (!update.isSnapshot || initialSnapshot) {
+                final ShiftAwareListener.Update downstream = new ShiftAwareListener.Update(
+                        update.rowsAdded.clone(), update.rowsRemoved.clone(), totalMods, update.shifted, modifiedColumnSet);
+                return (coalescer == null) ? new Index.IndexUpdateCoalescer(prevIndex, downstream) : coalescer.update(downstream);
+            } else {
+                return coalescer;
+            }
+        }
+    }
+
+    private void freeRows(final Index removedIndex) {
+        if (removedIndex.empty()) {
+            return;
+        }
+
+        final Index rowsToFree = removedIndex.intersect(populatedRows);
+        doFreeRows(rowsToFree);
+    }
+
+    private void processPendingFrees() {
+        if (pendingFree == null) {
+            return;
+        }
+        doFreeRows(pendingFree);
+        pendingFree.close();
+        pendingFree = null;
+    }
+
+    private void doFreeRows(final Index rowsToFree) {
+        final TLongArrayList freedRows = new TLongArrayList(rowsToFree.intSize());
```

Comment:
We should just use a `LongChunk<OrderedKeyIndices>` with the appropriate sort kernel and skip translating to an index. Or use a random builder if we think it might save space.